src/main/java/org/opensearch/geospatial/ip2geo/action/PutDatasourceRequest.java

```diff
@@ -0,0 +1,147 @@
+
+import org.opensearch.action.ActionRequestValidationException;
+import org.opensearch.action.support.master.AcknowledgedRequest;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.io.stream.StreamOutput;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.core.ParseField;
+import org.opensearch.core.xcontent.ObjectParser;
+import org.opensearch.geospatial.ip2geo.common.DatasourceManifest;
+
+/**
+ * GeoIP datasource creation request
+ */
+@Getter
+@Setter
+@Log4j2
+public class PutDatasourceRequest extends AcknowledgedRequest<PutDatasourceRequest> {
+    private static final ParseField ENDPOINT_FIELD = new ParseField("endpoint");
+    private static final ParseField UPDATE_INTERVAL_IN_DAYS_FIELD = new ParseField("update_interval_in_days");
+    /**
+     * @param datasourceName the datasource name
+     * @return the datasource name
+     */
+    private String datasourceName;
+    /**
+     * @param endpoint url to a manifest file for a datasource
+     * @return url to a manifest file for a datasource
+     */
+    private String endpoint;
+    /**
+     * @param updateIntervalInDays update interval of a datasource
+     * @return update interval of a datasource
+     */
+    private TimeValue updateIntervalInDays;
+
+    /**
+     * Parser of a datasource
+     */
+    public static final ObjectParser<PutDatasourceRequest, Void> PARSER;
+    static {
+        PARSER = new ObjectParser<>("put_datasource");
+        PARSER.declareString((request, val) -> request.setEndpoint(val), ENDPOINT_FIELD);
+        PARSER.declareLong((request, val) -> request.setUpdateIntervalInDays(TimeValue.timeValueDays(val)), UPDATE_INTERVAL_IN_DAYS_FIELD);
+    }
+
+    /**
+     * Default constructor
+     * @param datasourceName name of a datasource
+     */
+    public PutDatasourceRequest(final String datasourceName) {
+        this.datasourceName = datasourceName;
+    }
+
+    /**
+     * Constructor with stream input
+     * @param in the stream input
+     * @throws IOException IOException
+     */
+    public PutDatasourceRequest(final StreamInput in) throws IOException {
+        super(in);
+        this.datasourceName = in.readString();
+        this.endpoint = in.readString();
+        this.updateIntervalInDays = in.readTimeValue();
+    }
+
+    @Override
+    public void writeTo(final StreamOutput out) throws IOException {
+        super.writeTo(out);
+        out.writeString(datasourceName);
+        out.writeString(endpoint);
+        out.writeTimeValue(updateIntervalInDays);
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        ActionRequestValidationException errors = new ActionRequestValidationException();
+        validateEndpoint(errors);
+        validateUpdateInterval(errors);
+        return errors.validationErrors().isEmpty() ? null : errors;
+    }
+
+    private void validateEndpoint(final ActionRequestValidationException errors) {
```

Comment:
if we have these validation function then don't you think these validation functions should also be called during the setters?

Reply:
No need to. Only need to validate the value when it is passed from customer. Once it is validated and processed, the value don't get changed after that.